File: app/src/main/java/org/thoughtcrime/securesms/jobmanager/persistence/JobStorage.java
package org.thoughtcrime.securesms.jobmanager.persistence;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.WorkerThread;

import java.util.List;
import java.util.Set;

public interface JobStorage {

  @WorkerThread
  void init();

  @WorkerThread
  void insertJobs(@NonNull List<FullSpec> fullSpecs);

  @WorkerThread
  @Nullable JobSpec getJobSpec(@NonNull String id);

  @WorkerThread
  @NonNull List<JobSpec> getAllJobSpecs();

  @WorkerThread
  @NonNull List<JobSpec> getPendingJobsWithNoDependenciesInCreatedOrder(long currentTime);

  @WorkerThread
  @NonNull List<JobSpec> getJobsInQueue(@NonNull String queue);

  @WorkerThread
  int getJobCountForFactory(@NonNull String factoryKey);

  @WorkerThread
  int getJobCountForFactoryAndQueue(@NonNull String factoryKey, @NonNull String queueKey);

  @WorkerThread
  boolean areQueuesEmpty(@NonNull Set<String> queueKeys);

  @WorkerThread
  void updateJobRunningState(@NonNull String id, boolean isRunning);

  @WorkerThread
  void updateJobAfterRetry(@NonNull String id, boolean isRunning, int runAttempt, long nextRunAttemptTime, @Nullable byte[] serializedData);

  @WorkerThread
  void updateAllJobsToBePending();

  @WorkerThread
  void updateJobs(@NonNull List<JobSpec> jobSpecs);

  @WorkerThread
  void deleteJob(@NonNull String id);

  @WorkerThread
  void deleteJobs(@NonNull List<String> ids);

  @WorkerThread
  @NonNull List<ConstraintSpec> getConstraintSpecs(@NonNull String jobId);

  @WorkerThread
  @NonNull List<ConstraintSpec> getAllConstraintSpecs();

  @WorkerThread
  @NonNull List<DependencySpec> getDependencySpecsThatDependOnJob(@NonNull String jobSpecId);

  @WorkerThread
  @NonNull List<DependencySpec> getAllDependencySpecs();
}

File: app/src/main/java/org/thoughtcrime/securesms/jobmanager/persistence/JobStorage.kt
package org.thoughtcrime.securesms.jobmanager.persistence

import androidx.annotation.WorkerThread
import org.thoughtcrime.securesms.jobs.MinimalJobSpec
import java.util.function.Predicate

interface JobStorage {
  @WorkerThread
  fun init()

  @WorkerThread
  fun insertJobs(fullSpecs: List<FullSpec>)

  @WorkerThread
  fun getJobSpec(id: String): JobSpec?

  @WorkerThread
  fun getAllMatchingFilter(predicate: Predicate<JobSpec>): List<JobSpec>

  @WorkerThread
  fun getNextEligibleJob(currentTime: Long, filter: (MinimalJobSpec) -> Boolean): JobSpec?

  @WorkerThread
  fun getJobsInQueue(queue: String): List<JobSpec>

  @WorkerThread
  fun getJobCountForFactory(factoryKey: String): Int

  @WorkerThread
  fun getJobCountForFactoryAndQueue(factoryKey: String, queueKey: String): Int

  @WorkerThread
  fun areQueuesEmpty(queueKeys: Set<String>): Boolean

  @WorkerThread
  fun markJobAsRunning(id: String, currentTime: Long)

  @WorkerThread
  fun updateJobAfterRetry(id: String, currentTime: Long, runAttempt: Int, nextBackoffInterval: Long, serializedData: ByteArray?)

  @WorkerThread
  fun updateAllJobsToBePending()

  @WorkerThread
  fun updateJobs(jobSpecs: List<JobSpec>)

  @WorkerThread
  fun transformJobs(transformer: (JobSpec) -> JobSpec)

  @WorkerThread
  fun deleteJob(id: String)

  @WorkerThread
  fun deleteJobs(ids: List<String>)

  @WorkerThread
  fun getConstraintSpecs(jobId: String): List<ConstraintSpec>

  @WorkerThread
  fun getDependencySpecsThatDependOnJob(jobSpecId: String): List<DependencySpec>

  @WorkerThread
  fun debugGetJobSpecs(limit: Int): List<JobSpec>

  @WorkerThread
  fun debugGetConstraintSpecs(limit: Int): List<ConstraintSpec>

  @WorkerThread
  fun debugGetAllDependencySpecs(): List<DependencySpec>
}
